src/execution/sema/sema_builtin.cpp

```diff
@@ -1840,6 +1840,68 @@ void Sema::CheckMathTrigCall(ast::CallExpr *call, ast::Builtin builtin) {
   call->SetType(GetBuiltinType(return_kind));
 }
 
+void Sema::CheckAtomicCall(ast::CallExpr *call, ast::Builtin builtin) {
+  const auto &call_args = call->Arguments();
+
+  // Permissive arg check to protect the following argument dereference
+  if (!CheckArgCountAtLeast(call, 1)) return;
+
+  auto operand_type = call_args[0]->GetType()->GetPointeeType();
+  if (operand_type == nullptr) {
+    ReportIncorrectCallArg(call, 0, "'dest' must be pointer to an integral type");
+    return;
+  }
+
+  /* TODO(John): Ideally we want to support CMPXCHG on pointers as well, but the
```

Comment:
Nit: although I understand this issue, I'm not sure it's appropriate here. Compare-exchange on pointers and recursive data types are unrelated. Either state that compare-exchange on pointers is future work, or add support here now.

Reply:
Removed the comment and left the code as integral only.